apisix/plugins/datadog.lua

```diff
@@ -0,0 +1,269 @@
+--
+-- Licensed to the Apache Software Foundation (ASF) under one or more
+-- contributor license agreements.  See the NOTICE file distributed with
+-- this work for additional information regarding copyright ownership.
+-- The ASF licenses this file to You under the Apache License, Version 2.0
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+local core = require("apisix.core")
+local plugin = require("apisix.plugin")
+local batch_processor = require("apisix.utils.batch-processor")
+local fetch_log = require("apisix.utils.log-util").get_full_log
+local ngx = ngx
+local udp = ngx.socket.udp
+local format = string.format
+local concat = table.concat
+local buffers = {}
+local ipairs = ipairs
+local tostring = tostring
+local stale_timer_running = false
+local timer_at = ngx.timer.at
+
+local plugin_name = "datadog"
+local defaults = {
+    host = "127.0.0.1",
+    port = 8125,
+    namespace = "apisix",
+    constant_tags = {"source:apisix"}
+}
+
+local schema = {
+    type = "object",
+    properties = {
+        buffer_duration = {type = "integer", minimum = 1, default = 60},
+        inactive_timeout = {type = "integer", minimum = 1, default = 5},
+        batch_max_size = {type = "integer", minimum = 1, default = 5000},
+        max_retry_count = {type = "integer", minimum = 1, default = 1},
+    }
+}
+
+local metadata_schema = {
+    type = "object",
+    properties = {
+        host = {type = "string", default= defaults.host},
+        port = {type = "integer", minimum = 0, default = defaults.port},
+        namespace = {type = "string", default = defaults.namespace},
+        constant_tags = {
+            type = "array",
+            items = {type = "string"},
+            default = defaults.constant_tags
+        }
+    },
+}
+
+local _M = {
+    version = 0.1,
+    priority = 495,
+    name = plugin_name,
+    schema = schema,
+    metadata_schema = metadata_schema,
+}
+
+function _M.check_schema(conf, schema_type)
+    if schema_type == core.schema.TYPE_METADATA then
+        return core.schema.check(metadata_schema, conf)
+    end
+    return core.schema.check(schema, conf)
+end
+
+local function generate_tag(conf, const_tags)
+    local tags
+    if const_tags and #const_tags > 0 then
+        tags = core.table.deepcopy(const_tags)
+    else
+        tags = {}
+    end
+
+    if conf.route_id and conf.route_id ~= "" then
+        core.table.insert(tags, "route_id:" .. conf.route_id)
```

Comment:
```
idx = #tags + 1
...
tags[idx] = "route_id:"
tags[idx + 1] = conf.route_id
idx = idx + 2
```
might be faster. You can give it a try.

Reply:
>tags[idx] = "route_id:"
tags[idx + 1] = conf.route_id
idx = idx + 2



The required format is comma-separated key-value pair. where a particular key and value pair is separated by a colon ( `:` ). This idx + 2 won't be able to achieve that.

TBH, the insert is a perf improvement that `tags = tags .. ""blah"` as in this case the whole tags get copied again and again. This is not the case with table.insert("key:", val), it's just O(1) and the concat is O(N) so effectively O(N).

Asymptotically, both the idx approach that you suggested and the insert approach that we have currently is O(N). But I am not sure if insert has a little overhead than tags[idx]. Please let me know, I'll be happy to update it.

the updated approach will be
tags[idx] = "route_id:"  .. conf.route_id
idx = idx + 1
..

Reply:
The `table.insert` without idx will need to calculate the length of the table first (which is not an O(1) operation). The current behavior is fine as the optimization may be too trivial.
See:
https://github.com/openresty/luajit2/blob/cd2285f1b7792ccf61db37c10da7463b6b1a30a5/src/lib_table.c#L157
https://github.com/openresty/luajit2/blob/da4aecf2e79adaed6f6ec9726e330a77816efbe3/src/lj_tab.c#L663